### PR TITLE
Add MqttConnectOptions parameter to AWSIotMqttManager

### PR DESCRIPTION
**Pull request by dhartwich1991**
***Description of changes:***
Adds the option to add `MqttConnectOptions` to `AWSIotMqttManager`.

Hello, while using AWS IoT we noticed that there is no way to set `MqttConnectOptions` and we don't have access to setting parameters like the maximum inflight messages etc. 

We are using QOS1 and are running into a lot of `TOO MANY MESSAGES` issues:

```
...

org.eclipse.paho.client.mqttv3.internal.ClientState.send (ClientState.java:496)
org.eclipse.paho.client.mqttv3.internal.ClientComms.internalSend (ClientComms.java:132)
org.eclipse.paho.client.mqttv3.internal.ClientComms.sendNoWait (ClientComms.java:156)
org.eclipse.paho.client.mqttv3.MqttAsyncClient.publish (MqttAsyncClient.java:1027)
org.eclipse.paho.client.mqttv3.MqttAsyncClient.publish (MqttAsyncClient.java:991)
com.amazonaws.mobileconnectors.iot.AWSIotMqttManager.publishData (AWSIotMqttManager.java:1339)
io.moia.iot.IotConnection$publishData$2.invokeSuspend (IotConnection.kt:787)

...
```

This PR is not finished yet, but I would like to use it as a communication starter.
Are you in general open for such kind of changes? We really would like to avoid our private fork of this repo if possible!

Thank you! :)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by desokroshan** [comment]
@dhartwich1991 Thanks for submitting the PR. I will take a look and get back to you. Thanks!

File: aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotMqttManager.java [hunk 1]
```diff
@@ -827,6 +843,81 @@ private void connect(KeyStore keyStore, int portNumber, final AWSIotMqttClientSt
...
+     * using certificate and private key in keystore on the specified port.
+     *
+     * @param keyStore A keystore containing an keystore with a certificate and
+     *            private key. Use IotKeystoreHelper to get keystore.
+     * @param portNumber the client port, either 8883 or 443
+     * @param statusCallback When new MQTT session status is received the
+     *            function of callback will be called with new connection
+     *            status.
+     */
+    private void connect(KeyStore keyStore, int portNumber, final MqttConnectOptions connectOptions, final AWSIotMqttClientStatusCallback statusCallback) {
```
Comment (desokroshan) [inline aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotMqttManager.java]: This method is copy and paste of another overloaded method in this file. Can you please eliminate the redundancy?

File: aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotMqttManager.java [hunk 2]
```diff
@@ -910,6 +1001,90 @@ public void run() {
...
+    /**
+     * Initializes the MQTT session and connects to the specified MQTT server
+     * using AWS credentials.
+     *
+     * @param credentialsProvider AWS credentialsProvider used to create a WebSocket connection to AWS IoT.
+     * @param connectOptions MQTT connectOptions that enable custom parameters on the connection.
+     * @param statusCallback When new MQTT session status is received the function of callback will
+     *                       be called with new connection status.
+     */
+    public void connect(AWSCredentialsProvider credentialsProvider, final MqttConnectOptions connectOptions,
```
Comment (desokroshan) [inline aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotMqttManager.java]: Same as the last comment. Please remove the redundancy.

**Comment by desokroshan** [comment]
@dhartwich1991 Thanks for incorporating the suggested changes. I am discussing the PR internally within the team and am deliberating if adding `MqttConnectOptions` parameter which adds an additional layer of complexity to the connect API is the best route here.
To clarify my understanding, you are running into "TOO MANY MESSAGES" type errors because you are rapidly publishing messages with QoS being set to 1? Can you share some relevant code snippet which can be used to reproduce the issue? Better understanding of issue will be useful in coming up with alternative solutions.

**Comment by dhartwich1991** [comment]
Hey @desokroshan thanks for the review once again!

Yea we are running in to these `Too many publishes` errors because we are publishing a lot of updates (location of a vehicle (the app we're building is used inside a car) + other state of our devices). It can occur, that these updates stack up, because of bad connection (driving through a tunnel for example) and once we reconnect there are too many updates to be published.
To be 100% honest with you we are not exactly sure what is causing the `too many messages` problems but this is our most educated guess we can come up with, since the app is quite complex and state changes can trigger sending of new messages. Since most of these informations are important for our business we **need** to send them on QOS1

I would appreciate if we can find a solution!

Thank you and your team so much for your effort already!

**Comment by desokroshan** [comment]
Thanks for the details. Can share some code snippet which can be used to reproduce the issue?

**Comment by jamesonwilliams** [comment]
Looks like there was some discussion around next steps for this PR last year, but it has been lingering since then. Will close for now. Please re-open if there are updates to consider.

**Pull request closed without merging**